File: DS/Stacks/cpp/prefix_evaluation.cpp
#include <bits/stdc++.h>
using namespace std;

int prefixEvaluation(string s){
    stack<int> st;
    int a,b;

    for(int i=s.length()-1; i>=0;i--){
        if(isdigit(s[i])){
            st.push(s[i]-'0');
        }

        else{
            switch (s[i])
            {
            case '*':
                a = st.top();
                st.pop();
                b = st.top();
                st.pop();
                st.push(a*b);
                break;
            
            case '/':
                a = st.top();
                st.pop();
                b = st.top();
                st.pop();
                st.push(a/b);
                break;
            
            case '+':
                a = st.top();
                st.pop();
                b = st.top();
                st.pop();
                st.push(a+b);
                break;

            case '-':
                a = st.top();
                st.pop();
                b = st.top();
                st.pop();
                st.push(a-b);
                break;

            default:
                break;
            }
        }
    }
    
    return st.top();
}

int main(){
    string s = "-+*23*549";
    cout<<prefixEvaluation(s);
}

// output : 17

File: DS/Stacks/Python/prefix_evaluation.py
def prefix(str1, n):
    stack = []

    n -= 1

    while n >= 0:
        if str1[n].isnumeric():
            stack.append(int(str1[n]))

        else:
            if str1[n] == '+':
                op2 = stack.pop()
                op1 = stack.pop()
                stack.append(op2 + op1)

            elif str1[n] == '-':
                op2 = stack.pop()
                op1 = stack.pop()
                stack.append(op2 - op1)

            elif str1[n] == '*':
                op2 = stack.pop()
                op1 = stack.pop()
                stack.append(op2 * op1)

            elif str1[n] == '/':
                op2 = stack.pop()
                op1 = stack.pop()
                stack.append(op2 / op1)
        n -= 1

    return stack.pop()


s = "-+*23*549"
print(prefix(s, len(s)))

# output: 17